src/fsharp/FSharp.Compiler.Private/Utilities/ShardMap.fs

```diff
@@ -0,0 +1,1504 @@
+
+    ///let inline private bucketIndexOld (keyHash:int,bitdepth:int) = (keyHash &&& (~~~(-1 <<< (bitdepth)))) >>> 4// todo: improve substring bitmask calc
+
+    ///provides sub index in shards
+    let inline shardIndex (keyHash:int) = keyHash &&& 0b1111
+    let inline isNull v = Object.ReferenceEquals(null,v)
+
+    let inline getIndexes(key,bucketBitMask) = let hk  = key.GetHashCode() in bucketIndex(hk,bucketBitMask) , shardIndex(hk)
+    let inline higherRange (index:int,bitdepth:int) = (index ||| (1 <<< (bitdepth - 4)))
+
+    let inline newShardEmpty () = Array.zeroCreate<MapTree<'K,'V>>(ShardSize)
+    let inline newBucketEmpty bucketSize = Array.zeroCreate<Shard<'K,'V>>(bucketSize)
+
+    let inline eachMapInBucketClose (bucket:Bucket<'K,'V>,f:int -> int -> MapTree<'K,'V> -> unit,bucketCloseFn:int -> unit) =
+        let fo = OptimizedClosures.FSharpFunc<_,_,_,_>.Adapt(f)
+    
+        Tasks.Parallel.For(0, bucket.Length, fun bi ->
+            for si in 0 .. ShardSize - 1 do
+                let s = bucket.[bi]
+                if not( isNull s) then
+                    let m = s.[si]
+                    if not( isNull m) then
+                        fo.Invoke(bi,si,m)
+            bucketCloseFn bi
+        ) |> ignore
+
+    let inline eachMapInBucket (bucket:Bucket<'K,'V>,f:int -> int -> MapTree<'K,'V> -> unit) =
+        eachMapInBucketClose (bucket,f,ignore)
+
+    let getOrCreateShard (bucket:Bucket<'K,'V>,index:int) =
+        let mutable shrd = bucket.[index]
+        if isNull shrd then
+            shrd <- newShardEmpty ()
+            bucket.[index] <- shrd
+        shrd        
+
+    /////////////////////////////
+    /// Constructor helper
+
+
+    let init<'T,'K,'V when 'K : comparison>(counter:int, items : 'T seq,mapFn: 'T -> KeyValuePair<'K,'V>) =
+        let size = if counter < ShardSize then ShardSize else counter
+        let comparer = LanguagePrimitives.FastGenericComparer<'K>
+        let bitdepth = (calcBitMaskDepth size)
+        let bucketSize = bucketSizeFromBitDepth (bitdepth)
+        let bucketBitMask = calcSubBitMask bitdepth
+        let newBucket = Array.zeroCreate<Shard<'K,'V>>(bucketSize)
+
+        items
+        |> Seq.iter (fun item ->
+            let kvp = mapFn item
+            // let kh = kvp.Key.GetHashCode()
+            // let bi = bucketIndex(kh,bucketBitMask)
+            // let si = shardIndex(kh)
+            let bi, si = getIndexes(kvp.Key,bucketBitMask)
+            let shrd = getOrCreateShard(newBucket,bi)
+            let m = shrd.[si]
+            shrd.[si] <- if isNull m then MapOne (kvp.Key,kvp.Value) else MapTree.add comparer kvp.Key kvp.Value m
+            )
+        (counter,newBucket)
+
+open util
+open MapTree
+open System.Threading.Tasks
+
+    /// Shard Map
+    ////////////////////////////
+
+type ShardMap<'K,'V  when 'K : equality and 'K : comparison >(icount:int, nBucket:Shard<'K,'V> []) =
```

Comment:
``nBucket`` feels like an odd name for an array?

Reply:
Apologies, naming conventions not set, my hash space array is made up of a jagged array so I called the internal  immutable slices "Shards", and the parent array holding the shards a Bucket (although the hash conflict MapTrees are more like traditional buckets). "nBucket" is just the temp naming of a "new Bucket" of shards to maintain immutability, happy to change this (and bucket naming convention for any actual merge).